ppl/cmd/zar/ls/command.go

```diff
@@ -54,69 +59,101 @@ func (c *Command) Run(args []string) error {
 		})
 	}
-	return archive.Walk(context.TODO(), ark, func(chunk chunk.Chunk) error {
-		c.printDir(ark, "", chunk.ZarDir(), pattern)
+	return archive.Walk(context.TODO(), c.ark, func(chunk chunk.Chunk) error {
+		c.printChunk(0, chunk, pattern)
 		return nil
 	})
 }
 
-func fileExists(path iosrc.URI) bool {
-	info, err := iosrc.Stat(context.TODO(), path)
-	if err != nil {
-		return false
-	}
-	if fsinfo, ok := info.(os.FileInfo); ok {
-		return !fsinfo.IsDir()
+func (c *Command) printSpanInfo(si archive.SpanInfo, pattern string) {
+	fmt.Println(si.Range(c.ark.DataOrder) + ":")
+	for _, chunk := range si.Chunks {
+		c.printChunk(1, chunk, pattern)
 	}
-	return true
 }
 
-func (c *Command) printDir(ark *archive.Archive, rangeStr string, dir iosrc.URI, pattern string) {
-	if pattern != "" {
-		path := dir.AppendPath(pattern)
-		if fileExists(path) {
-			fmt.Println(c.printable(ark, rangeStr, dir, pattern))
-		}
+func (c *Command) printChunk(indent indent, chunk chunk.Chunk, pattern string) {
+	str := c.chunkString(chunk)
+	var children string
+	if !c.lflag {
+		fmt.Println(indent.String() + str)
 		return
 	}
-	if !c.lflag {
-		fmt.Println(c.printable(ark, rangeStr, dir, ""))
-	} else {
-		entries, err := iosrc.ReadDir(context.TODO(), dir)
-		if err != nil {
-			fmt.Fprintf(os.Stderr, "error listing directory: %v", err)
-			return
+	str += "/"
+	children = c.indicesString(str, chunk)
+	children += c.mapsString(str, chunk, pattern)
+	fmt.Print(children)
+}
+
+func (c *Command) indicesString(prefix string, chunk chunk.Chunk) string {
+	indices, err := index.Indices(context.TODO(), chunk.ZarDir(), c.defs)
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "error listing directory: %v", err)
+		return ""
+	}
+	var b strings.Builder
+	for _, i := range indices {
+		var str string
+		if c.indexDesc {
+			str = i.Definition.String()
+		} else {
+			str = i.Filename()
 		}
-		for _, e := range entries {
-			fmt.Println(c.printable(ark, rangeStr, dir, e.Name()))
+		fmt.Fprintln(&b, prefix+str)
+	}
+	return b.String()
+}
+
+func (c *Command) mapsString(prefix string, chunk chunk.Chunk, pattern string) string {
+	files, err := index.ListFilenames(context.TODO(), chunk.ZarDir())
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "error listing directory: %v", err)
+		return ""
+	}
+	var b strings.Builder
+	for _, file := range files {
+		if pattern == "" || pattern == file {
+			fmt.Fprintln(&b, prefix+file)
 		}
 	}
+	return b.String()
 }
 
-func (c *Command) printable(ark *archive.Archive, rangeStr string, zardir iosrc.URI, objPath string) string {
+func (c *Command) chunkString(chunk chunk.Chunk) string {
 	if c.showRanges {
-		return path.Join(rangeStr, objPath)
+		return chunk.Range()
 	}
 	if c.relativePaths {
-		return strings.TrimSuffix(ark.DataPath.RelPath(zardir.AppendPath(objPath)), "/")
+		return strings.TrimSuffix(c.ark.DataPath.RelPath(chunk.ZarDir()), "/")
 	}
-	return zardir.AppendPath(objPath).String()
+	return chunk.ZarDir().String()
+}
+
+type indent int
```

Comment:
For simplicity, I'd remove this and just use a boolean when needed.

I do like & appreciate the resulting output with the grouped spaninfos!